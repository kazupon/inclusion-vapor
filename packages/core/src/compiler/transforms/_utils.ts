import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
