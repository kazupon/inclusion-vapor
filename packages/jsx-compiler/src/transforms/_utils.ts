import { parse } from '@babel/parser'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, JSXElement, JSXFragment } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const {
      body: [statement]
    } = parse(source, {
      sourceType: 'module',
      plugins: ['jsx']
    }).program
    let children!: JSXElement[] | JSXFragment['children']
    if (statement.type === 'ExpressionStatement') {
      children =
        statement.expression.type === 'JSXFragment'
          ? statement.expression.children
          : (statement.expression.type === 'JSXElement'
            ? [statement.expression]
            : [])
    }

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children,
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
