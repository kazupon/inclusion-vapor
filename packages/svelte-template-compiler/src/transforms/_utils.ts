import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, RootIRNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (
    source: string,
    overrideOptions: CompilerOptions = {}
  ): {
    ast: RootNode
    ir: RootIRNode
    code: string
    helpers: Set<string>
    vaporHelpers: Set<string>
  } => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
