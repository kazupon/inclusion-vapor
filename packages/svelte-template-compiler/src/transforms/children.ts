// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformChildren.ts

import { DynamicFlag, IRNodeTypes, isSvelteText } from '../ir/index.ts'
import { transformNode } from '../transform.ts'

import type { IRDynamicInfo, SvelteElement } from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT ||
    node.type === 'Fragment' ||
    node.type === 'InlineComponent' ||
    node.type === 'Slot'
  if (node.type !== 'Element' && !isFragment) {
    return
  }
  processChildren(node, context, isFragment)
}

export function processChildren(
  node: Parameters<NodeTransform>[0],
  context: Parameters<NodeTransform>[1],
  isFragment: boolean
): ReturnType<NodeTransform> {
  const children = node.children || []

  // normalize svelte text
  for (const [index, child] of [...children].entries()) {
    if (isSvelteText(child) && !child.data.trim()) {
      child.data = ' '
      if (!index) {
        children.splice(0, 1)
      } else if (index === children.length) {
        children.splice(-1, 1)
      }
    }
  }

  for (const [i, child] of children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<SvelteElement>)
  }
}

function processDynamicChildren(context: TransformContext<SvelteElement>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length > 0) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`

          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())

          context.registerOperation({
            type: IRNodeTypes.INSERT_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference(),
            anchor
          })
        } else {
          context.registerOperation({
            type: IRNodeTypes.PREPEND_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference()
          })
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length > 0) {
    context.registerOperation({
      type: IRNodeTypes.INSERT_NODE,
      elements: prevDynamics.map(child => child.id!),
      parent: context.reference()
    })
  }
}
