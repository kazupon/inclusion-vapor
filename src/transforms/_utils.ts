import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type { RootIRNode as VaporRootIRNode } from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      prefixIdentifiers: true,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      prefixIdentifiers: true,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
