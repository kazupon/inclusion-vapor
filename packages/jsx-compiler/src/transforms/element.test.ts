import { BindingTypes, NodeTypes } from '@vue-vapor/compiler-dom'
import { compile as vaporCompile } from '@vue-vapor/compiler-vapor'
import { describe, expect, test } from 'vitest'
import { IRNodeTypes } from '../ir/index.ts'
import { makeCompile } from './_utils.ts'
import { transformChildren } from './children.ts'
import { transformElement } from './element.ts'
import { transformText } from './text.ts'

import type { BindingMetadata } from '@vue-vapor/compiler-dom'

const compileWithElementTransform = makeCompile({
  nodeTransforms: [transformElement, transformChildren, transformText]
})

test('native elements', () => {
  const vaporSource = `
<div class="container">
  <div class="header">
    <p style="color: red;">Hello</p>
    <img src="foo.jpg" width="500" height="600" />
    <form action="/submit" method="post">
      <input type="text" />
      <input type="submit" />
    </form>
  </div>
</div>
`
  const jsxSource = `(<>${vaporSource}</>)`
  const { ir: _, code, vaporHelpers } = compileWithElementTransform(jsxSource)
  const expectedResult = vaporCompile(vaporSource)
  expect(code).toMatchSnapshot('jsx')
  expect(expectedResult.code).toMatchSnapshot('vapor')
  // NOTE:
  // There are differences in the handling around spaces and line breaks between vue compiler and jsx compiler.
  // about details, see the snapshot
  // expect(code).toEqual(expectedResult.code)
  expect(vaporHelpers).toEqual(expectedResult.vaporHelpers)
})

describe('component', () => {
  test('basic', () => {
    const vaporSource = `
<div class="container">
  <Foo class="foo"></Foo>
</div>
`
    const jsxSource = `(<>${vaporSource}</>)`
    const { ir: _, code, vaporHelpers } = compileWithElementTransform(jsxSource)
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(code).toEqual(expectedResult.code)
    expect(vaporHelpers).toEqual(expectedResult.vaporHelpers)
  })

  test('import + resolve component', () => {
    const vaporSource = `<Foo/>`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code, ir, vaporHelpers } = compileWithElementTransform(jsxSource)
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(code).toEqual(expectedResult.code)
    expect(vaporHelpers).contains.all.keys('resolveComponent', 'createComponent')
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        id: 0,
        tag: 'Foo',
        asset: true,
        root: true,
        props: [[]]
      }
    ])
  })

  // TODO: WIP, how should we handle components?
  test('resolve component from setup bindings', () => {
    const vaporSource = `<Example />`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code, ir, vaporHelpers } = compileWithElementTransform(jsxSource, {
      prefixIdentifiers: false,
      bindingMetadata: {
        Example: BindingTypes.SETUP_MAYBE_REF
      }
    })
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(expectedResult.code).toMatchSnapshot('vapor')
    // expect(code).toEqual(expectedResult.code)
    expect(vaporHelpers).not.toContain('resolveComponent')
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        tag: 'Example',
        asset: false
      }
    ])
  })

  // TODO: WIP, how should we handle components?
  test('resolve component from setup bindings (inline)', () => {
    const vaporSource = `<Example />`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code, vaporHelpers } = compileWithElementTransform(jsxSource, {
      inline: true,
      bindingMetadata: {
        Example: BindingTypes.SETUP_MAYBE_REF
      }
    })
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(code).contains(`unref(Example)`)
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(vaporHelpers).not.toContain('resolveComponent')
    expect(vaporHelpers).toContain('unref')
  })

  // TODO: WIP, how should we handle components?
  test('resolve component from setup bindings (inline const)', () => {
    const vaporSource = `<Example />`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code, vaporHelpers } = compileWithElementTransform(jsxSource, {
      inline: true,
      bindingMetadata: {
        Example: BindingTypes.SETUP_CONST
      }
    })
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(vaporHelpers).not.toContain('resolveComponent')
  })

  // TODO: WIP, how should we handle components?
  test('resolve namespaced component from setup bindings', () => {
    const vaporSource = `<Foo.Example />`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code, vaporHelpers } = compileWithElementTransform(jsxSource, {
      bindingMetadata: {
        Foo: BindingTypes.SETUP_MAYBE_REF
      }
    })
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(code).contains(`_ctx.Foo.Example`)
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(vaporHelpers).not.toContain('resolveComponent')
  })

  // TODO: WIP, how should we handle components?
  test('resolve namespaced component from setup bindings (inline const)', () => {
    const vaporSource = `<Foo.Example />`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code, vaporHelpers } = compileWithElementTransform(jsxSource, {
      inline: true,
      bindingMetadata: {
        Foo: BindingTypes.SETUP_CONST
      }
    })
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(code).contains(`Foo.Example`)
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(vaporHelpers).not.toContain('resolveComponent')
  })

  // TODO: WIP, how should we handle components?
  test('resolve namespaced component from props bindings (inline)', () => {
    const vaporSource = `<Foo.Example />`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code, vaporHelpers } = compileWithElementTransform(jsxSource, {
      inline: true,
      bindingMetadata: {
        Foo: BindingTypes.PROPS
      }
    })
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(code).contains(`Foo.Example`)
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(vaporHelpers).not.toContain('resolveComponent')
  })

  // TODO: WIP, how should we handle components?
  test('resolve namespaced component from props bindings (non-inline)', () => {
    const vaporSource = `<Foo.Example />`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code, vaporHelpers } = compileWithElementTransform(jsxSource, {
      inline: false,
      bindingMetadata: {
        Foo: BindingTypes.PROPS
      }
    })
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(code).contains('_ctx.Foo.Example')
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(vaporHelpers).not.toContain('resolveComponent')
  })

  // TODO: WIP, how should we handle components?
  test('do not resolve component from non-script-setup bindings', () => {
    const bindingMetadata: BindingMetadata = {
      Example: BindingTypes.SETUP_MAYBE_REF
    }
    Object.defineProperty(bindingMetadata, '__isScriptSetup', {
      value: false
    })
    const vaporSource = `<Example />`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code, ir, vaporHelpers } = compileWithElementTransform(jsxSource, {
      bindingMetadata
    })
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(vaporHelpers).toContain('resolveComponent')
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        id: 0,
        tag: 'Example',
        asset: true
      }
    ])
  })

  // TODO: WIP, how should we handle components?
  test('generate single root component', () => {
    const vaporSource = `<Comp/>`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code } = compileWithElementTransform(jsxSource, {
      bindingMetadata: { Comp: BindingTypes.SETUP_CONST }
    })
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(code).contains('_createComponent(_ctx.Comp, null, null, true)')
  })

  // TODO: WIP, how should we handle components?
  test('generate multi root component', () => {
    const vaporSource = `<Comp/>123`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code } = compileWithElementTransform(jsxSource, {
      bindingMetadata: { Comp: BindingTypes.SETUP_CONST }
    })
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(code).contains('_createComponent(_ctx.Comp)')
  })

  test('static props', () => {
    const vaporSource = `
<Foo id="foo" class="bar" />
`
    const jsxSource = `(<>${vaporSource}</>)`
    const { code, ir } = compileWithElementTransform(jsxSource)
    const expectedResult = vaporCompile(vaporSource)
    expect(code).toMatchSnapshot('jsx')
    expect(expectedResult.code).toMatchSnapshot('vapor')
    expect(code).toEqual(expectedResult.code)

    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        tag: 'Foo',
        asset: true,
        root: true,
        props: [
          [
            {
              key: {
                type: NodeTypes.SIMPLE_EXPRESSION,
                content: 'id',
                isStatic: true
              },
              values: [
                {
                  type: NodeTypes.SIMPLE_EXPRESSION,
                  content: 'foo',
                  isStatic: true
                }
              ]
            },
            {
              key: {
                type: NodeTypes.SIMPLE_EXPRESSION,
                content: 'class',
                isStatic: true
              },
              values: [
                {
                  type: NodeTypes.SIMPLE_EXPRESSION,
                  content: 'bar',
                  isStatic: true
                }
              ]
            }
          ]
        ]
      }
    ])
  })
})

test('static props', () => {
  const vaporSource = `<div id="foo" class="bar" />`
  const jsxSource = `(<>${vaporSource}</>)`
  const { code, ir } = compileWithElementTransform(jsxSource)
  const template = '<div id="foo" class="bar"></div>'
  expect(code).toMatchSnapshot()
  expect(code).contains(JSON.stringify(template))
  expect(ir.template).toMatchObject([template])
  expect(ir.block.effect).lengthOf(0)
})

test('props + children', () => {
  const vaporSource = `<div id="foo"><span/></div>`
  const jsxSource = `(<>${vaporSource}</>)`
  const { code, ir } = compileWithElementTransform(jsxSource)
  const template = '<div id="foo"><span></span></div>'
  expect(code).toMatchSnapshot()
  expect(code).contains(JSON.stringify(template))
  expect(ir.template).toMatchObject([template])
  expect(ir.block.effect).lengthOf(0)
})

test.skip('invalid html nesting', () => {
  const source = `
<p><div>123</div></p>
<form><form/></form>
`
  const { ir, code, vaporHelpers: _ } = compileWithElementTransform(source)
  const expectedResult = vaporCompile(source)
  expect(code).toMatchSnapshot('received')
  expect(expectedResult.code).toMatchSnapshot('expected')
  expect(code).toEqual(expectedResult.code)

  expect(ir.template).toEqual(['<div>123</div>', '<p></p>', '<form></form>'])
  expect(ir.block.dynamic).toMatchObject({
    children: [
      { id: 1, template: 1, children: [{ id: 0, template: 0 }] },
      { id: 3, template: 2, children: [{ id: 2, template: 2 }] }
    ]
  })

  expect(ir.block.operation).toMatchObject([
    { type: IRNodeTypes.INSERT_NODE, parent: 1, elements: [0] },
    { type: IRNodeTypes.INSERT_NODE, parent: 3, elements: [2] }
  ])
})

test('empty template', () => {
  const { code } = compileWithElementTransform('<></>')
  expect(code).toMatchSnapshot()
  expect(code).contain('return null')
})
