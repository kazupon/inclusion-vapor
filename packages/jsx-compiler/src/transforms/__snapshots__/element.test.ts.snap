// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`component > basic > jsx 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent, insert as _insert, template as _template } from 'vue/vapor';
const t0 = _template("<div class=\\"container\\"></div>")

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n1 = t0()
  const n0 = _createComponent(_component_Foo, [
    { class: () => ("foo") }
  ])
  _insert(n0, n1)
  return n1
}"
`;

exports[`component > basic > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent, insert as _insert, setInheritAttrs as _setInheritAttrs, template as _template } from 'vue/vapor';
const t0 = _template("<div class=\\"container\\"></div>")

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n1 = t0()
  const n0 = _createComponent(_component_Foo, [
    { class: () => ("foo") }
  ])
  _insert(n0, n1)
  _setInheritAttrs(false)
  return n1
}"
`;

exports[`component > do not resolve component from non-script-setup bindings > jsx 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx, $props, $emit, $attrs, $slots) {
  const _component_Example = _resolveComponent("Example")
  const n0 = _createComponent(_component_Example, null, null, true)
  return n0
}"
`;

exports[`component > do not resolve component from non-script-setup bindings > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Example = _resolveComponent("Example")
  const n0 = _createComponent(_component_Example, null, null, true)
  return n0
}"
`;

exports[`component > generate multi root component > jsx 1`] = `
"import { createComponent as _createComponent, template as _template } from 'vue/vapor';
const t0 = _template("123")

export function render(_ctx, $props, $emit, $attrs, $slots) {
  const n1 = t0()
  const n0 = _createComponent(_ctx.Comp)
  return [n0, n1]
}"
`;

exports[`component > generate multi root component > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent, template as _template } from 'vue/vapor';
const t0 = _template("123")

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n1 = t0()
  const n0 = _createComponent(_component_Comp)
  return [n0, n1]
}"
`;

exports[`component > generate single root component > jsx 1`] = `
"import { createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx, $props, $emit, $attrs, $slots) {
  const n0 = _createComponent(_ctx.Comp, null, null, true)
  return n0
}"
`;

exports[`component > generate single root component > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n0 = _createComponent(_component_Comp, null, null, true)
  return n0
}"
`;

exports[`component > import + resolve component > jsx 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n0 = _createComponent(_component_Foo, null, null, true)
  return n0
}"
`;

exports[`component > import + resolve component > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n0 = _createComponent(_component_Foo, null, null, true)
  return n0
}"
`;

exports[`component > resolve component from setup bindings (inline const) > jsx 1`] = `
"((_ctx) => {
  const n0 = _createComponent(Example, null, null, true)
  return n0
})()"
`;

exports[`component > resolve component from setup bindings (inline const) > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Example = _resolveComponent("Example")
  const n0 = _createComponent(_component_Example, null, null, true)
  return n0
}"
`;

exports[`component > resolve component from setup bindings (inline) > jsx 1`] = `
"((_ctx) => {
  const n0 = _createComponent(_unref(Example), null, null, true)
  return n0
})()"
`;

exports[`component > resolve component from setup bindings (inline) > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Example = _resolveComponent("Example")
  const n0 = _createComponent(_component_Example, null, null, true)
  return n0
}"
`;

exports[`component > resolve component from setup bindings > jsx 1`] = `
"import { createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx, $props, $emit, $attrs, $slots) {
  const n0 = _createComponent(Example, null, null, true)
  return n0
}"
`;

exports[`component > resolve component from setup bindings > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Example = _resolveComponent("Example")
  const n0 = _createComponent(_component_Example, null, null, true)
  return n0
}"
`;

exports[`component > resolve namespaced component from props bindings (inline) > jsx 1`] = `
"((_ctx) => {
  const n0 = _createComponent(Foo.Example, null, null, true)
  return n0
})()"
`;

exports[`component > resolve namespaced component from props bindings (inline) > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Foo46Example = _resolveComponent("Foo.Example")
  const n0 = _createComponent(_component_Foo46Example, null, null, true)
  return n0
}"
`;

exports[`component > resolve namespaced component from props bindings (non-inline) > jsx 1`] = `
"import { createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx, $props, $emit, $attrs, $slots) {
  const n0 = _createComponent(_ctx.Foo.Example, null, null, true)
  return n0
}"
`;

exports[`component > resolve namespaced component from props bindings (non-inline) > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Foo46Example = _resolveComponent("Foo.Example")
  const n0 = _createComponent(_component_Foo46Example, null, null, true)
  return n0
}"
`;

exports[`component > resolve namespaced component from setup bindings (inline const) > jsx 1`] = `
"((_ctx) => {
  const n0 = _createComponent(Foo.Example, null, null, true)
  return n0
})()"
`;

exports[`component > resolve namespaced component from setup bindings (inline const) > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Foo46Example = _resolveComponent("Foo.Example")
  const n0 = _createComponent(_component_Foo46Example, null, null, true)
  return n0
}"
`;

exports[`component > resolve namespaced component from setup bindings > jsx 1`] = `
"import { createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx, $props, $emit, $attrs, $slots) {
  const n0 = _createComponent(_ctx.Foo.Example, null, null, true)
  return n0
}"
`;

exports[`component > resolve namespaced component from setup bindings > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Foo46Example = _resolveComponent("Foo.Example")
  const n0 = _createComponent(_component_Foo46Example, null, null, true)
  return n0
}"
`;

exports[`component > static props > jsx 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n0 = _createComponent(_component_Foo, [
    {
      id: () => ("foo"), 
      class: () => ("bar")
    }
  ], null, true)
  return n0
}"
`;

exports[`component > static props > vapor 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n0 = _createComponent(_component_Foo, [
    {
      id: () => ("foo"), 
      class: () => ("bar")
    }
  ], null, true)
  return n0
}"
`;

exports[`empty template 1`] = `
"
export function render(_ctx) {
  return null
}"
`;

exports[`native elements > jsx 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div class=\\"container\\"><div class=\\"header\\"><p style=\\"color: red;\\">Hello</p> <img src=\\"foo.jpg\\" width=\\"500\\" height=\\"600\\"> <form action=\\"/submit\\" method=\\"post\\"><input type=\\"text\\"> <input type=\\"submit\\"></form></div></div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`native elements > vapor 1`] = `
"import { setInheritAttrs as _setInheritAttrs, template as _template } from 'vue/vapor';
const t0 = _template("<div class=\\"container\\"><div class=\\"header\\"><p style=\\"color: red;\\">Hello</p><img src=\\"foo.jpg\\" width=\\"500\\" height=\\"600\\"><form action=\\"/submit\\" method=\\"post\\"><input type=\\"text\\"><input type=\\"submit\\"></form></div></div>")

export function render(_ctx) {
  const n0 = t0()
  _setInheritAttrs(false)
  return n0
}"
`;

exports[`props + children 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div id=\\"foo\\"><span></span></div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`static props 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div id=\\"foo\\" class=\\"bar\\"></div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;
